### PR TITLE
API

### PR DESCRIPTION
**Pull request by sroberts**
So I had to recreate the PR & Branch, but here's the start of my REST API. I could probably use some help with creating new configs and config groups using this Null Pattern.

### CODE REVIEW

File: server.rb [hunk 1]
```diff
@@ -126,34 +126,139 @@ def difftime(oldtime, newtime)
...
-    # This next line is necessary because osqueryd does not send the
-    # enroll_secret as a POST param.
-    begin
-      json_data = JSON.parse(request.body.read)
-      params.merge!(json_data)
-    rescue
+  namespace '/configurations' do
+    # CRUD CONFIG
+
+    post '/new' do
```
Comment (ukd1) [inline server.rb]: wouldn't it be better to be ``post '/'``?

File: server.rb [hunk 2]
```diff
@@ -126,34 +126,139 @@ def difftime(oldtime, newtime)
...
-  post '/config' do
-    # This next line is necessary because osqueryd does not send the
-    # enroll_secret as a POST param.
-    begin
-      params.merge!(JSON.parse(request.body.read))
-    rescue
+  namespace '/configuration_groups' do
+    # CRUD CONFIG_GROUPS
+
+    post '/new' do
```
Comment (ukd1) [inline server.rb]: ditto

File: server.rb [hunk 3]
```diff
@@ -126,34 +126,139 @@ def difftime(oldtime, newtime)
...
+      end
     end
-    logdebug "value in node_key is #{params['node_key']}"
-    client = GuaranteedEndpoint.find_by node_key: params['node_key']
-    logdebug "Received endpoint: #{client.inspect}"
-    client.get_config user_agent: request.user_agent
   end
 
+  namespace '/endpoints' do
+    post do
```
Comment (ukd1) [inline server.rb]: this is inconsistent with the above post methods (even though this does nothing)

**Comment by sroberts** [comment]
@ukd1 Thanks for the thoughts. Still very WIP (as you noted).

**Comment by blackfist** [comment]
So I went to one of my rails projects and ran `rake routes` and what @ukd1 says is true. If we're going to be consistent with the way rest is done in other frameworks, then the `GET` request that brings up a form should be at `/thing/new` but the post goes to `/thing`. Same for api. Obviously no get request for a form in the api, but the `POST` should go to `/configuration_groups` and `/configurations`.

That means I've also got some fixing to do on the html stuff too. Damn.

I'll start helping with this api PR after BSidesDFW

**Comment by sroberts** [comment]
So I need to tweak something to create `configurations` correctly. Right now they keep getting rolled back from the DB and don't have the `config_json`, `version`, & `notes` fields set correctly.

Then I just need to add [basic authentication](http://recipes.sinatrarb.com/p/middleware/rack_auth_basic_and_digest).

**Comment by sroberts** [comment]
So I included two different solutions to posting configurations.
- `/api/configurations/new`: Which requires an body option specifying a Configuration Group.
- `/api/configuration_groups/2/configurations/new`: This may be slightly "RESTful" but sorta uglier IMHO. 

I'm sorta up in the air about which one I prefer. 

Needs authentication and some validity checks, but otherwise the API piece is getting pretty good. I also need to add @ukd1's updates as well. But I'd love to get general feedback in the mean time.

**Comment by blackfist** [comment]
@sroberts it looks like you moved the /config path under the /enroll namespace. Was that intentional? It broke some of the tests.

**Comment by sroberts** [comment]
That must have been a mistake, that absolutely should not be.

**Comment by blackfist** [comment]
ok, I'll fix it on my end.

**Comment by sroberts** [comment]
So this is looking really good. I need to add a [helper method for auth](http://stackoverflow.com/questions/3479737/sinatra-api-authentication). Otherwise this looks pretty good, don't you think?

**Comment by sroberts** [comment]
Double check me on this. My merge tool might have messed something up, it left some of the API namespace in the server.rb.

**Comment by blackfist** [comment]
Looks like we have the `/auth` stuff both in server.rb and also in `lib/controllers/auth.rb`. I think you should strip it out of server because I want the stuff in `lib/controllers/auth.rb` to the authoritative.

**Comment by sroberts** [comment]
Fixed. I knew I missed something.

**Comment by sroberts** [comment]
Testing this out now, but I believe there's a bug. When no API keys are defined going to the "/apikeys" page results in an error:

```
16:24:53 web.1  | [41165] 127.0.0.1 - - [02/Dec/2015:16:24:53 -0500] "GET /apikeys HTTP/1.1" 500 30 0.0270
16:25:04 web.1  | D, [2015-12-02T16:25:04.209279 #41165] DEBUG -- :   APIKey Load (0.7ms)  SELECT "api_keys".* FROM "api_keys"
16:25:04 web.1  | E, [2015-12-02T16:25:04.210110 #41165] ERROR -- : SQLite3::SQLException: no such table: api_keys: SELECT "api_keys".* FROM "api_keys"
16:25:04 web.1  | 2015-12-02 16:25:04 - ActiveRecord::StatementInvalid - SQLite3::SQLException: no such table: api_keys: SELECT "api_keys".* FROM "api_keys":
16:25:04 web.1  |       /Users/sroberts/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:91:in `initialize'
16:25:04 web.1  |       /Users/sroberts/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:91:in `new'
```

Seems like we need some handling so it defaults to create a new key if no keys exist.

**Comment by blackfist** [comment]
Have you run "rake db:migrate" to make sure the table exists?

**Comment by sroberts** [comment]
@blackfist Sorry, edited my comment after I read the freakin' docs. My bad.

**Comment by sroberts** [comment]
After a long an arduous path it looks like the API stuff is all ready to go. There's still some little things I'd like to add, but nothing that should block this getting merged.

**Pull request merged**